source/inputoutput.rst

```diff
@@ -0,0 +1,114 @@
+
+.. note:: In case the quantities you'd like to output are fairly simple, you could use Scala's ``println`` function to directly output what you need.
+
+Saving your output to a CSV file
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+BharatSim relies on a trait called ``SimulationListener`` to help output data .
+
+``SimulationListener`` contains 4 methods, each of which allow us to perform a task in one of the following situations:
+
+* At the start of the simulation
+* At the start of every time step
+* At the end of every time step
+* At the end of the simulation
+
+The BharatSim engine also contains a class called ``CsvOutputGenerator``, an extension of ``SimulationListener`` which has two attributes:
+
+* ``path``, the desired path for the output file to be stored
+* ``csvSpecs``, a user-defined class that outputs the headers and the rows required. Note that this user-defined class should extend the ``CSVSpecs`` trait and override the ``getHeaders`` and ``getRows`` methods.
+
+This class writes the headers at the start of the simulation, writes the rows at the start of every time step, and closes the writer at the end of the simulation.
+
+Output at a single instant of time
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+We can define a class as follows:
+
+.. code-block:: scala
+
+  import com.bharatsim.engine.Context
+  import com.bharatsim.engine.listeners.CSVSpecs
+
+  class MyOutputSpec(context: Context) extends CSVSpecs {
+    override def getHeaders: List[String] =
+      List(
+        "Header1",
+        "Header2",
+        "Header3"
+      )
+    override def getRows(): List[List[Any]] = {
+      val elementInRow: String = "row" + context.getCurrentStep.toString
+      val row = List(
+        elementInRow,
+        elementInRow,
+        elementInRow
+      )
+      List(row)
+    }
+  }
+
+Now, we need to create an instance of the ``CsvOutputGenerator`` class that uses ``MyOutputSpec``, and call the required methods. We do that by adding the following code snippet inside ``simulation.defineSimulation`` in the main function:
+
+.. code-block:: scala
+
+  var outputGenerator = new CsvOutputGenerator("src/main/resources/output.csv", new MyOutputSpec(context))
+  outputGenerator.onSimulationStart(context)
+  outputGenerator.onStepStart(context)
+  outputGenerator.onSimulationEnd(context)
+
+.. note:: Calling the ``onStepEnd`` method of the class isn't necessary, as the ``CsvOutputGenerator`` class currently does nothing when it's called.
+
+Remember to import ``com.bharatsim.engine.listeners.CsvOutputGenerator``!
+
+.. hint:: In case you want your outputs generated *after* the simulation is completed, you can place the above 4 lines of code inside ``simulation.onCompleteSimulation``.
+
```

Comment:
Put in a screenshot of what the output CSV file will look like.

Reply:
adding a code-block instead with the contents of the csv